### PR TITLE
feat: Checkpoint pipeline infrastructure (Phase 1)

### LINKED ISSUE #493: Checkpoint Pipeline Infrastructure (Phase 1)
## Overview
Implement the core pipeline infrastructure for the new checkpoint system architecture - Phase 1 of the 5-PR implementation plan.

## Parent Issue
Part of #248 - Checkpoint System Refactor

### Three-Layer Pipeline Design
The checkpoint system is being refactored into three distinct, composable layers that work together through a pipeline pattern:

1. **Checkpoint Acquisition Layer** (PR #458)
   - Responsible for fetching checkpoint files from various sources (local, S3, HTTP, etc.)
   - Returns a local path to the checkpoint file
   - Handles caching, retries, and network failures

2. **Loading Orchestration Layer** (New PR - Phase 2)
   - Implements different strategies for loading checkpoints into models
   - Strategies include: weights-only, transfer learning, warm start, cold start
   - Manages optimizer and scheduler state restoration

3. **Model Transformation Layer** (PR #410)
   - Applies post-loading modifications to the model
   - Includes freezing layers, adding LoRA adapters, quantization
   - Executed after checkpoint loading is complete

### This Issue's Role
This issue tracks the **foundation layer** that enables all three layers above to work together:

```
┌─────────────────────────────────────────────────┐
│   User Configuration (Hydra YAML)               │
└─────────────────────────────────────────────────┘
                    ↓
┌─────────────────────────────────────────────────┐
│   Pipeline Infrastructure (THIS ISSUE)          │
│   - CheckpointContext (carries state)           │
│   - PipelineStage (base abstraction)            │
│   - CheckpointPipeline (orchestrator)           │
│   - ComponentRegistry (factory pattern)         │
└─────────────────────────────────────────────────┘
                    ↓
    Orchestrates the following stages in order:
                    ↓
    1. Acquisition → 2. Loading → 3. Transformation
```

### Data Flow Example
```python
# 1. User provides configuration
config = {
    'checkpoint': {
        'source': {'type': 's3', 'bucket': 'models', 'key': 'model.ckpt'},
        'loading': {'type': 'transfer_learning', 'skip_mismatched': True}
    },
    'model_modifier': {
        'modifiers': [{'type': 'freeze', 'layers': ['encoder']}]
    }
}

# 2. Pipeline creates context
context = CheckpointContext(model=model, config=config)

# 3. Pipeline executes stages
context = await acquisition_stage.process(context)  # Downloads from S3
context = await loading_stage.process(context)      # Loads with transfer learning
context = await modifier_stage.process(context)     # Freezes encoder

# 4. Result: modified model with checkpoint loaded
model = context.model
```

### 1. Core Abstractions (`training/src/anemoi/training/checkpoint/base.py`)
```python
@dataclass
class CheckpointContext:
    """Carries state through pipeline stages"""
    checkpoint_path: Optional[Path] = None
    checkpoint_data: Optional[Dict] = None
    model: Optional[nn.Module] = None
    optimizer: Optional[Optimizer] = None
    scheduler: Optional[Any] = None
    metadata: Dict[str, Any] = field(default_factory=dict)
    config: Optional[DictConfig] = None
    
class PipelineStage(ABC):
    """Base class for all pipeline stages"""
    @abstractmethod
    async def process(self, context: CheckpointContext) -> CheckpointContext:
        pass
```

### 2. Pipeline Orchestrator (`training/src/anemoi/training/checkpoint/pipeline.py`)
```python
class CheckpointPipeline:
    """Orchestrates checkpoint processing through stages"""
    
    def __init__(self, stages: List[PipelineStage]):
        self.stages = stages
    
    async def execute(self, initial_context: CheckpointContext) -> CheckpointContext:
        context = initial_context
        for stage in self.stages:
            context = await stage.process(context)
        return context
```

### 3. Registry Pattern (`training/src/anemoi/training/checkpoint/registry.py`)
- Component registration for sources, loaders, and modifiers
- Factory methods for creating components from config
- Extensibility for adding new component types

### 4. Error Handling (`training/src/anemoi/training/checkpoint/exceptions.py`)
- `CheckpointError` - Base exception
- `CheckpointNotFoundError` - File not found
- `CheckpointLoadError` - Loading failures
- `CheckpointIncompatibleError` - Compatibility issues

## Implementation Checklist
- [x] Core abstractions defined
- [x] Pipeline orchestrator implemented
- [x] Registry pattern established
- [x] Error handling comprehensive
- [x] Async utilities implemented
- [x] Unit tests complete (70% coverage target)
- [ ] Integration tests complete
- [ ] Documentation complete

### 1. Async-First Architecture
- All I/O operations are async for better performance
- Support for parallel downloads from multiple sources
- Non-blocking pipeline execution

### 2. Context Pattern
- `CheckpointContext` carries all state through pipeline
- Immutable context updates for debugging
- Clear data flow through stages

### 3. Registry Pattern
- Extensible component registration
- Factory methods for instantiation
- Plugin-style architecture for new components

### 4. Error Handling Strategy
- Fail-fast with recovery mechanisms
- Comprehensive error types for different failures
- Retry logic with exponential backoff

### Unit Tests (`tests/checkpoint/test_pipeline.py`)
- Test pipeline execution order
- Test context passing between stages
- Test error propagation
- Mock all external dependencies

### Integration Tests (`tests/checkpoint/test_integration.py`)
- Test full pipeline with mock stages
- Test async execution
- Test error recovery
- Test registry functionality

## Migration Path
This infrastructure will support legacy configurations through adapters:
- `load_weights_only=True` → WeightsOnlyLoader
- `transfer_learning=True` → TransferLearningLoader
- `resume_from_checkpoint` → WarmStartLoader

## Related Issues
- Parent: #248 (Checkpoint System Refactor)
- Blocks: Loading Orchestration issue
- Related to: #458 (Checkpoint Acquisition), #410 (Model Modifiers)

## Branch
- Working branch: `feature/checkpoint-pipeline-infrastructure`

## Success Criteria
- [ ] All tests passing
- [ ] No breaking changes to existing functionality (yet)
- [ ] Documentation complete
- [ ] Ready for Phase 2 development to begin

### PR DESCRIPTION
**Pull request by JesperDramsch**
## Summary

This PR implements Phase 1 of the checkpoint pipeline infrastructure, establishing the core abstractions and pipeline pattern for flexible checkpoint handling in Anemoi.

## Changes

- **Core abstractions**: Added  for state management and  base class for all pipeline stages
- **Pipeline orchestration**: Implemented  with support for async/sync execution and stage management
- **Dynamic component discovery**: Created  with automatic discovery of checkpoint sources, loaders, and modifiers using module inspection
- **Comprehensive error handling**: Added exception hierarchy for checkpoint operations
- **Utility functions**: Implemented helpers for download retry, validation, and metadata extraction

## Technical Details

The component catalog uses a hybrid approach for identifying abstract classes:
- ABC inheritance detection
- Abstract method checking
- Name-based convention (Base* prefix)

This ensures true dynamic discovery without hardcoded component lists, making the system easily extensible.

## Testing

- Unit tests for all core components
- Integration tests for pipeline execution
- Tests for dynamic component discovery with various abstract class patterns

## Related Issues

Closes #493 

## Next Steps

This is Phase 1 of a multi-phase implementation:
- Phase 2: Three parallel layers (Checkpoint Acquisition #458, Loading Orchestration #494, Model Transformation #410)
- Phase 3: System integration and migration (#495)

----
📚 Documentation preview 📚: https://anemoi-training--501.org.readthedocs.build/en/501/

----
📚 Documentation preview 📚: https://anemoi-graphs--501.org.readthedocs.build/en/501/

----
📚 Documentation preview 📚: https://anemoi-models--501.org.readthedocs.build/en/501/

### CODE REVIEW

**Comment by JesperDramsch** [comment]
# Phase 1: Checkpoint Pipeline Infrastructure - Progress Update

## Summary

This PR implements the foundational infrastructure for the new checkpoint architecture, establishing the core abstractions, pipeline pattern, and utilities that all checkpoint operations will build upon.

**Related Issues**: #493 (Phase 1), Part of 5-phase checkpoint architecture refactoring
**Branch**: `feature/checkpoint-pipeline-infrastructure`
**Status**: Core complete, configuration integration pending review

## 🎯 Motivation

The current checkpoint handling in anemoi-training is monolithic and difficult to extend. This PR establishes a clean, three-layer pipeline architecture:

```
┌─────────────────────────────────────────────────┐
│        Model Transformation Layer               │
│         (Post-loading modifications)            │
├─────────────────────────────────────────────────┤
│         Loading Orchestration Layer             │
│    (Strategies for applying checkpoints)        │
├─────────────────────────────────────────────────┤
│        Checkpoint Acquisition Layer             │
│      (Obtaining checkpoint from sources)        │
├─────────────────────────────────────────────────┤
│         Pipeline Infrastructure  ← THIS PR      │
│         (Core abstractions & context)           │
└─────────────────────────────────────────────────┘
```

## 📦 Changes Included

### New Files (VERIFIED)
```
training/src/anemoi/training/checkpoint/
├── __init__.py                    # Package initialization (2.6KB)
├── base.py                        # Core abstractions (16KB)
├── pipeline.py                    # Pipeline orchestrator (26KB)
├── catalog.py                     # Component discovery (22KB)
├── exceptions.py                  # Exception hierarchy (19KB)
├── formats.py                     # Multi-format support (19KB)
└── utils.py                       # Async utilities (17KB)

training/tests/checkpoint/
├── conftest.py                    # Test fixtures (9.7KB)
├── test_base.py                   # Core tests (8.1KB)
├── test_pipeline.py               # Pipeline tests (10KB)
├── test_catalog.py                # Catalog tests (11KB)
├── test_formats.py                # Format tests (26KB)
├── test_utils.py                  # Utility tests (37KB)
└── test_exceptions.py             # Exception tests (32KB)
```

### Modified Files (VERIFIED)
```
training/pyproject.toml            # Added optional checkpoint dependency (line 67)
```

### Pending Additions (Not Yet in PR)
```
training/src/anemoi/training/schemas/training.py   # TODO: Add CheckpointPipelineConfig
config/training/checkpoint_pipeline/*.yaml         # TODO: Create config templates
```

## 🔑 Key Features

### 1. Pipeline Pattern
Clean, composable stages for checkpoint processing:
```python
# Example usage (programmatic)
pipeline = CheckpointPipeline(
    stages=[
        CheckpointSource(...),      # Phase 2
        LoadingStrategy(...),       # Phase 2
        ModelModifier(...),         # Phase 2
    ],
    async_execution=True
)

context = await pipeline.execute(initial_context)
```

### 2. Component Catalog
Automatic discovery of pipeline components using reflection:
```python
# No manual registration needed!
ComponentCatalog.list_sources()    # Auto-discovers CheckpointSource subclasses
ComponentCatalog.list_loaders()    # Auto-discovers LoadingStrategy subclasses
ComponentCatalog.list_modifiers()  # Auto-discovers ModelModifier subclasses
```

### 3. Multi-Format Support
Seamless handling of different checkpoint formats:
- PyTorch Lightning checkpoints
- Standard PyTorch checkpoints
- SafeTensors format (optional)
- Raw state_dict formats

Auto-detection and conversion included.

### 4. Comprehensive Error Handling
Rich exception hierarchy for debugging:
```python
try:
    context = await pipeline.execute(context)
except CheckpointNotFoundError as e:
    logger.error(f"Checkpoint not found: {e}")
except CheckpointIncompatibleError as e:
    logger.error(f"Incompatible checkpoint: {e}")
```

### 5. Async-First Design
Efficient async operations with sync compatibility:
```python
# Async mode (default)
context = await pipeline.execute(context)

# Sync mode (when needed)
pipeline = CheckpointPipeline(stages, async_execution=False)
context = pipeline.execute(context)
```

## 🧪 Testing Strategy

### Coverage Metrics (VERIFIED)
- **Total Tests**: 213
- **Passing**: 211 (99.1%)
- **Skipped**: 2
- **Code Coverage**: 80% (798/1003 lines)

### Test Execution
```bash
# Run all checkpoint tests
pytest training/tests/checkpoint/ -v

# Run with coverage
pytest training/tests/checkpoint/ --cov=anemoi.training.checkpoint --cov-report=html

# Quick verification
pytest training/tests/checkpoint/ -q
# Output: 211 passed, 2 skipped in 46.46s
```

### Test Coverage by Module
- `base.py`: Core abstractions
- `pipeline.py`: Pipeline orchestration with Hydra
- `catalog.py`: Component discovery
- `exceptions.py`: Error hierarchy
- `formats.py`: Multi-format detection/conversion
- `utils.py`: Async downloads, validation, checksums

## 🔄 Integration Points

### With PR #464 (Checkpoint Acquisition)
```python
# PR #464 will implement CheckpointSource subclasses
class S3Source(PipelineStage):  # Uses base.py abstractions
    async def process(self, context: CheckpointContext) -> CheckpointContext:
        # Load from S3, populate context.checkpoint_data
        ...
```

### With PR #494 (Loading Orchestration)
```python
# PR #494 will implement LoadingStrategy subclasses
class TransferLearningLoader(PipelineStage):  # Uses base.py abstractions
    async def process(self, context: CheckpointContext) -> CheckpointContext:
        # Apply checkpoint to model with flexibility
        ...
```

### With PR #442 (Model Modifiers)
```python
# PR #442 will implement ModelModifier as PipelineStage
class FreezingModifier(PipelineStage):  # Uses base.py abstractions
    async def process(self, context: CheckpointContext) -> CheckpointContext:
        # Freeze model layers post-loading
        ...
```

## ⚠️ Breaking Changes

**None**. This PR is pure infrastructure - no existing code is modified or broken.

## 📋 Configuration Schema

**Status**: ⚠️ **Not yet integrated** - needs to be added before merge

Planned addition to `training/src/anemoi/training/schemas/training.py`:
```python
class CheckpointPipelineConfig(BaseModel):
    """Configuration for checkpoint pipeline"""
    stages: List[Dict[str, Any]] = Field(default_factory=list)
    async_execution: bool = Field(default=True)
    cache_dir: Optional[str] = Field(default=None)
    max_retries: int = Field(default=3)
    timeout: int = Field(default=300)
```

## 🔍 Review Checklist

### Code Quality ✅
- [x] All code follows project style guidelines (ruff)
- [x] Type hints on all functions (mypy validated)
- [x] Comprehensive docstrings (NumPy style)
- [x] No code duplication

### Testing ✅
- [x] Unit tests for all new functionality
- [x] 99.1% test pass rate (211/213)
- [x] 80% overall test coverage achieved
- [x] All critical paths covered

### Documentation ⚠️
- [x] Public API documented in code
- [ ] **User guide needed**
- [ ] **Migration guide needed**
- [x] Architecture diagrams included

### Configuration ⚠️
- [ ] **Schema integration pending**
- [ ] **Config templates pending**
- [x] Optional dependencies properly marked
- [x] Graceful degradation when optional deps missing

### Compatibility ✅
- [x] No breaking changes to existing API
- [x] Backward compatible
- [x] Optional dependencies properly marked
- [x] Graceful degradation

## 🚀 Deployment Notes

### Dependencies
**Required**: Already in `pyproject.toml`
- `torch >= 2.2.0`
- `omegaconf`
- `hydra-core`

**Optional**: Added in this PR (pyproject.toml line 67)
```toml
[project.optional-dependencies]
checkpoint = ["safetensors>=0.4.0"]
```

### Migration Path
No migration needed - this is new infrastructure. Existing checkpoint code remains unchanged.

## 📊 Performance Impact

- **Memory**: Minimal overhead (~1-2MB for pipeline infrastructure)
- **Speed**: Async operations enable concurrent checkpoint processing
- **CPU**: No significant impact, efficient component discovery

## 🔜 Next Steps

### Before Merge (Recommended)
1. **Add configuration schema** to `training/src/anemoi/training/schemas/training.py`
2. **Create config templates** in `config/training/checkpoint_pipeline/`
3. **Add basic user documentation**

### After Merge
1. **Phase 2 PRs** can immediately build on this:
   - PR #464: Checkpoint Acquisition (in progress)
   - PR #494: Loading Orchestration (ready to start)
   - PR #442: Model Transformation (in progress)

2. **Documentation**: Comprehensive user guides

3. **Phase 3**: Integration & migration utilities

## 📝 Honest Assessment

### What's Strong
- Core pipeline architecture is solid and tested
- Component discovery eliminates boilerplate
- Multi-format support is comprehensive
- Error handling is thorough
- Test coverage is good (80%, 99.1% pass rate)
- All abstractions are in place for Phase 2

### What Needs Work
- Configuration schema not integrated yet
- User documentation is minimal
- Config templates don't exist
- Some edge cases not covered in tests (20% uncovered)

### Production Readiness
- **For developers building Phase 2**: ✅ Ready now
- **For end users**: ⚠️ Needs config integration and docs

---

**Reviewers**: Please focus on:
1. API design of `CheckpointContext` and `PipelineStage`
2. Component discovery mechanism in `catalog.py`
3. Exception hierarchy completeness
4. Whether to require config schema integration before merge

File: training/src/anemoi/training/train/train.py [hunk 1]
```diff
@@ -167,6 +167,66 @@ def model(self) -> GraphForecaster:
...
+        if hasattr(self.config.training, "checkpoint_pipeline") and self.config.training.checkpoint_pipeline:
+            return self._use_modern_checkpoint_pipeline(model)
+
+        # Use legacy checkpoint logic for backward compatibility
+        return self._use_legacy_checkpoint_logic(model, **kwargs)
+
+    def _use_modern_checkpoint_pipeline(self, model: GraphForecaster) -> GraphForecaster:
+        """Use the modern checkpoint pipeline system."""
+        try:
+            import asyncio
```
Comment (anaprietonem) [inline training/src/anemoi/training/train/train.py]: is the asyncio functionality optional? (as in could people use the pipeline without this?)

File: training/src/anemoi/training/utils/checkpoint.py [hunk 2]
```diff
@@ -10,11 +10,17 @@
...
 
 import logging
+import pickle  # noqa: S403  # pickle required for PyTorch checkpoint loading
 from pathlib import Path
 
 import torch
 import torch.nn as nn
 
+# Import pipeline exceptions for consistent error handling
+from anemoi.training.checkpoint.exceptions import CheckpointConfigError
```
Comment (anaprietonem) [inline training/src/anemoi/training/utils/checkpoint.py]: @bdvllrs also implemented some exception https://github.com/ecmwf/anemoi-core/blob/main/models/src/anemoi/models/migrations/migrator.py for the migration component. I some might attack different aspects but I wonder if we should just have a common script to store all of this and avoid duplications.

File: training/src/anemoi/training/utils/checkpoint.py [hunk 3]
```diff
@@ -71,44 +77,178 @@ def save_inference_checkpoint(model: torch.nn.Module, metadata: dict, save_path:
...
+        "transfer_learning_loading() is deprecated and will be removed in a future version. "
+        "Use the checkpoint pipeline system with TransferLearningLoader instead. "
+        "See documentation for migration guide.",
+        DeprecationWarning,
+        stacklevel=2,
+    )
+
+    # Load the checkpoint with proper error handling
+    # Get device from model parameters, defaulting to CPU
+    device = next(model.parameters()).device if len(list(model.parameters())) > 0 else "cpu"
```
Comment (anaprietonem) [inline training/src/anemoi/training/utils/checkpoint.py]: could we add a log to show what device is being used?

File: training/src/anemoi/training/checkpoint/catalog.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+from __future__ import annotations
+
+import importlib
+import inspect
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class ComponentCatalog:
```
Comment (anaprietonem) [inline training/src/anemoi/training/checkpoint/catalog.py:56]: Seems like useful functionality but I wonder if this should live in anemoi-utils or potentially https://github.com/ecmwf/anemoi-registry/tree/main?

(as in how much of this is for classes related to checkpoints or extensive to any other anemoi class?)

File: training/src/anemoi/training/checkpoint/formats.py [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+        "Supported extensions: .ckpt, .pt, .pth, .bin, .safetensors",
+        extension,
+        path,
+    )
+    return "lightning"
+
+
+def load_checkpoint(
+    checkpoint_path: Path | str,
+    checkpoint_format: Literal["lightning", "pytorch", "safetensors", "state_dict"] | None = None,
```
Comment (anaprietonem) [inline training/src/anemoi/training/checkpoint/formats.py]: Can we use these formats at the moment? (as in safetensors is not yet supported, I would be in favour of keeping this simple for now and extending this when safetensors functionality is introduced)!

File: training/src/anemoi/training/checkpoint/catalog.py (hunk 4 above)
Comment (anaprietonem) [reply training/src/anemoi/training/checkpoint/catalog.py:56]: Also it could be interesting to get @gareth-j and @jjlk takes as if this could be relevant in the context or the anemoi-catalogue?

File: training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    import pytorch_lightning as pl
+    from omegaconf import DictConfig
+
+    from anemoi.training.checkpoint.base import CheckpointContext
+    from anemoi.training.checkpoint.pipeline import CheckpointPipeline
+
+LOGGER = logging.getLogger(__name__)
+
+
+class CheckpointPipelineCallback(Callback):
```
Comment (anaprietonem) [inline training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py]: Do I get correctly that the way the pipeline is then orchestrated and executed is via a Pytorch Lightning callback?

File: training/src/anemoi/training/checkpoint/formats.py [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+    if checkpoint_format is None:
+        checkpoint_format = detect_checkpoint_format(path)
+
+    try:
+        if checkpoint_format == "safetensors":
+            if not HAS_SAFETENSORS:
+                msg = (
+                    f"Cannot load safetensors checkpoint '{path}': safetensors library not available.\n"
+                    "Install with: pip install safetensors\n"
+                    "Or use a different checkpoint format (.ckpt, .pt, .pth)"
```
Comment (anaprietonem) [inline training/src/anemoi/training/checkpoint/formats.py]: What would be the benefit of introducing these many formats here? and how easy it would then to share checkpoints between users with many formats? (guess related to my comment that I think we could consider safetensors in a second phase)

File: training/src/anemoi/training/checkpoint/catalog.py [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+>>> from anemoi.training.checkpoint.catalog import ComponentCatalog
+>>>
+>>> # List dynamically discovered components
+>>> print(ComponentCatalog.list_sources())
+>>> ['local', 's3', 'http', 'gcs', 'azure']
+>>>
+>>> # Get component target path
+>>> target = ComponentCatalog.get_source_target('s3')
+>>> print(target)
+>>> 'anemoi.training.checkpoint.sources.S3Source'
```
Comment (anaprietonem) [inline training/src/anemoi/training/checkpoint/catalog.py:35]: in the case of s3 bucket - how would authentication be handled? I'd focus for now in supporting local checkpoints.

File: training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+if TYPE_CHECKING:
+    import pytorch_lightning as pl
+    from omegaconf import DictConfig
+
+    from anemoi.training.checkpoint.base import CheckpointContext
+    from anemoi.training.checkpoint.pipeline import CheckpointPipeline
+
+LOGGER = logging.getLogger(__name__)
+
+
```
Comment (sahahner) [inline training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py]: Could this file possibly be unified with this file? [callbacks/checkpoint.py](https://github.com/ecmwf/anemoi-core/blob/main/training/src/anemoi/training/diagnostics/callbacks/checkpoint.py)

File: training/src/anemoi/training/checkpoint/catalog.py (hunk 4 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/checkpoint/catalog.py:56]: @gareth-j @jjlk - I'd love your input on whether this component discovery pattern
would be useful in anemoi-registry/anemoi-catalogue context.

Current approach:
- ComponentCatalog discovers subclasses of training-specific base classes
- Uses module inspection to find implementations automatically
- Eliminates need for manual registration (but gives a tool to find different options)

Questions:
1. Do we use similar discovery patterns?
2. Would a shared discovery utility in anemoi-utils be beneficial?
3. Any concerns about this approach from a registry/catalogue perspective?

For Phase 1, I'd suggest we're keeping this in training since it's tightly coupled to checkpoint
pipeline base classes. Happy to discuss extraction for another PR though.

File: training/src/anemoi/training/train/train.py (hunk 1 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/train/train.py]: Should be possible to use without and only needed if remote content is fetched.

File: training/src/anemoi/training/checkpoint/formats.py (hunk 5 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/checkpoint/formats.py]: Done

File: training/src/anemoi/training/checkpoint/formats.py (hunk 7 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/checkpoint/formats.py]: Default is still pytorch lightning, but we already have different checkpoints for different usecases.

File: training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py (hunk 9 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py]: Possibly, right now we have this separation:

| Aspect | checkpoint.py (AnemoiCheckpoint) | checkpoint_pipeline.py (CheckpointPipelineCallback) |
|--------|----------------------------------|-----------------------------------------------------|
| Purpose | **Saving** checkpoints | **Loading** checkpoints via pipeline |
| Extends | ModelCheckpoint | Callback |
| Hooks | `on_train_end`, `on_save_checkpoint` | `on_fit_start` |
| When | During/after training | Before training starts |

Merging these into the same file could be confusing as they "technically handle checkpoints" but have very different purposes.

File: training/src/anemoi/training/checkpoint/catalog.py (hunk 8 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/checkpoint/catalog.py:35]: Cloud authentication will be handled in Phase 2 (PR #458) following standard SDK patterns:
- S3: boto3 credentials (env vars, ~/.aws/credentials, IAM roles)
- GCS: google-cloud-storage default credentials
- Azure: azure-identity DefaultAzureCredential

For Phase 1, I'm focusing on local file support as you suggested. The pipeline supports passing
auth configuration via config when remote sources are added.

File: training/src/anemoi/training/utils/checkpoint.py (hunk 3 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/utils/checkpoint.py]: Sure thing.

File: training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py (hunk 6 above)
Comment (JesperDramsch) [reply training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py]: The checkpoint pipeline supports **three execution patterns**:

#### Pattern 1: Standalone Programmatic (Async)
```python
from anemoi.training.checkpoint import CheckpointPipeline, CheckpointContext

# Build pipeline
pipeline = CheckpointPipeline(stages=[...])

# Execute asynchronously
context = CheckpointContext(model=model)
result = await pipeline.execute(context)
```

#### Pattern 2: Standalone Programmatic (Sync)
```python
# Sync mode for simpler use cases
pipeline = CheckpointPipeline(stages=[...], async_execution=False)
result = pipeline.execute(context)  # No await needed
```

#### Pattern 3: PyTorch Lightning Callback
```python
from anemoi.training.diagnostics.callbacks import CheckpointPipelineCallback

# Via trainer configuration
trainer = pl.Trainer(
    callbacks=[CheckpointPipelineCallback(config)]
)

# Pipeline executes automatically at on_fit_start
trainer.fit(model)
```

File: training/src/anemoi/training/train/train.py (hunk 1 above)
Comment (anaprietonem) [reply training/src/anemoi/training/train/train.py]: this would be part of the next PR where the interface to control this pipeline will be defined right? (as in I can't see any config file here where aysncio is just enabled for remote fetching?)

File: training/src/anemoi/training/utils/checkpoint.py (hunk 3 above)
Comment (anaprietonem) [reply training/src/anemoi/training/utils/checkpoint.py]: sorry if might I be missing it, but has this being added?

File: training/src/anemoi/training/checkpoint/catalog.py (hunk 8 above)
Comment (anaprietonem) [reply training/src/anemoi/training/checkpoint/catalog.py:35]: Sounds good thanks for the clarification Jesper. I would suggest to delete any references to 'gcs' and 'azure' for now so we can properly discuss this for phase in the next phase.

File: training/src/anemoi/training/checkpoint/catalog.py (hunk 4 above)
Comment (anaprietonem) [reply training/src/anemoi/training/checkpoint/catalog.py:56]: Thanks Jesper! @gareth-j  @jjlk  it could be good to discuss about this at some point. I am happy to keep this here for now. But I think the class itself has some overlap with functionality already existing in utils 

https://github.com/ecmwf/anemoi-utils/blob/0a781151eadda721a107a8ce0499fcf0bea8441a/src/anemoi/utils/registry.py#L11
@floriankrb  @HCookie - what do you think?

Jesper I think this be kept like this for this phase but in the future I would like to understand how we can unify this functionality better so it doesn't diverge or duplicate. I would also include a short section on the docs about this before merging.

File: training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py (hunk 6 above)
Comment (anaprietonem) [reply training/src/anemoi/training/diagnostics/callbacks/checkpoint_pipeline.py]: For Pattern 1 and 2, would those be called before trainer.fit? or how would one coordinate their execution in parallel with the trainer? I think you mentioned these patterns will be documented so happy to wait for those docs if you think it will clarify my question

File: training/pyproject.toml [hunk 10]
```diff
@@ -87,9 +87,13 @@ optional-dependencies.profile = [
   "rich>=13.6",
   "tabulate>=0.9",
 ]
+optional-dependencies.remote = [
+  "aiohttp>=3.8",
```
Comment (anaprietonem) [inline training/pyproject.toml:98]: Just to flag this dependency is already introduced with Pytorch-geometric https://github.com/pyg-team/pytorch_geometric/blob/master/pyproject.toml. Do we import it directly in the checkpoint pipeline to also add it here?

File: training/docs/checkpoint_integration.rst [hunk 11]
```diff
@@ -0,0 +1,256 @@
+.. _checkpoint_integration:
+
+#################################
+ Checkpoint Pipeline Integration
```
Comment (anaprietonem) [inline training/docs/checkpoint_integration.rst:4]: Nice docs!

File: training/pyproject.toml (hunk 10 above)
Comment (JesperDramsch) [reply training/pyproject.toml:98]: Yes in utils
